astropy/coordinates/sky_coordinate.py

```diff
@@ -1132,6 +1132,10 @@ def is_equivalent_frame(self, other):
                 "Tried to do is_equivalent_frame on something that isn't frame-like"
             )
 
+    def _prepare_for_separation(self, other):
```

Comment:
How about calling this `_to_own_frame(self, other)`? I think that is a little more descriptive and possibly we need it elsewhere (e.g., in `position_angle` it looks like we're doing something similar). It is also nicely analogous with `Quantity._to_own_unit`). 

That said, I'm actually not sure this shouldn't live on `baseframe`... See general comment.

Reply:
Falling back to `BaseCoordinateFrame.separation()` doesn't always work:
```python
>>> from astropy.coordinates import SkyCoord
>>> sc1 = SkyCoord(1, 2, unit="deg", frame="fk5")
>>> sc2 = SkyCoord(1, 2, unit="deg", frame="fk5", equinox="B1950")
>>> sc1.separation(sc2)
<Angle 0.69815121 deg>
>>> sc1.frame.separation(sc2)
<Angle 0. deg>
```

See also https://github.com/astropy/astropy/blob/6b36e53c2608c43a35a5561122688fb9e04879bf/astropy/coordinates/tests/test_sky_coord.py#L892

But the suggestion about the name of the new private method is a good one.

Reply:
OK, slightly weird that `frame.separation` ignores `equinox` but let that be for another time. For now, good to know that we need to override.

Reply:
Hmm, the reason for the difference is because a `SkyCoord` is being passed to `BaseCoordinateFrame.separation()`.  `SkyCoord.separation()` explicitly blocks the merging-attributes functionality of `SkyCoord.transform_to()`, but `BaseCoordinateFrame.separation()` unsurprisingly does not have such a block since `BaseCoordinateFrame.transform_to()` simply has no merging-attributes functionality.  So, when a `SkyCoord` is passed to `BaseCoordinateFrame.separation()`, the "wrong" `transform_to()` method gets called, frame attributes are merged, and `sc1` is treated as if it has the same equinox as `sc2` because `sc1`'s equinox has not been set.  If you explicitly set the equinox for `sc1`, the answer is correct:
```python
>>> from astropy.coordinates import SkyCoord
>>> sc1 = SkyCoord(1, 2, unit="deg", frame="fk5", equinox="J2000")
>>> sc2 = SkyCoord(1, 2, unit="deg", frame="fk5", equinox="B1950")
>>> sc1.separation(sc2)
<Angle 0.69815121 deg>
>>> sc1.frame.separation(sc2)
<Angle 0.69815121 deg>
```
```

Reply:
This is the problematic line:
https://github.com/astropy/astropy/blob/6b36e53c2608c43a35a5561122688fb9e04879bf/astropy/coordinates/baseframe.py#L1801

Reply:
That seems like a good solution; I don't think it was ever meant to be the case that a `SkyCoord` with no explicit equinox set should be interpreted as one that can have arbitrary equinox.

Reply:
Following the suggestion of @ayshih it is indeed possible to remove the `SkyCoord` separation methods entirely without any loss of functionality. However, in that case these methods will not appear in the `SkyCoord` API documentation page. I would say that's an acceptable tradeoff, but I'd like this to be confirmed before I push the new commit. Note that in the narrative docs it is possible to replace links to the `SkyCoord` methods with links to the `BaseCoordinateFrame` methods, so there's no problem there.

Reply:
I'm OK with that - it seems important that the methods behave the same way, and the easiest way to ensure that is the case is to have just one!